src/compiler/checker.ts

```diff
@@ -8172,9 +8187,26 @@ namespace ts {
         }
 
         function filterType(type: Type, f: (t: Type) => boolean): Type {
-            return type.flags & TypeFlags.Union ?
-                getUnionType(filter((<UnionType>type).types, f)) :
-                f(type) ? type : neverType;
+            if (!(type.flags & TypeFlags.Union)) {

```

Comment:
Yes, it is an optimization to avoid doing any work if the filter function returns true for all constituents. But now that you mention it, it would make more sense to push the optimization into the `filter` function itself. Similar to how `concatenate` doesn't allocate a new array unless it has to, so could `filter` just return the argument array when the predicate is true for all elements.